File: widgetssdk/src/main/java/com/glia/widgets/chat/adapter/UploadAttachmentAdapter.java
package com.glia.widgets.chat.adapter;

import android.content.ContentResolver;
import android.content.Context;
import android.database.Cursor;
import android.provider.OpenableColumns;
import android.text.format.Formatter;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.glia.widgets.R;
import com.glia.widgets.fileupload.model.FileAttachment;
import com.glia.widgets.helper.Utils;
import com.google.android.material.progressindicator.LinearProgressIndicator;
import com.squareup.picasso.Picasso;

public class UploadAttachmentAdapter extends ListAdapter<FileAttachment, UploadAttachmentAdapter.ViewHolder> {
    public UploadAttachmentAdapter() {
        super(new DiffUtil.ItemCallback<FileAttachment>() {
            @Override
            public boolean areItemsTheSame(@NonNull FileAttachment oldItem, @NonNull FileAttachment newItem) {
                return oldItem.getUri().equals(newItem.getUri());
            }

            @Override
            public boolean areContentsTheSame(@NonNull FileAttachment oldItem, @NonNull FileAttachment newItem) {
                return oldItem.isReadyToSend() == newItem.isReadyToSend() &&
                        oldItem.getAttachmentStatus().equals(newItem.getAttachmentStatus());
            }
        });
    }

    private ItemCallback callback;

    public void setItemCallback(ItemCallback callback) {
        this.callback = callback;
    }

    @Override
    public int getItemViewType(int position) {
        return 0;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.chat_attachment_uploaded_item, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.onBind(getItem(position), callback);
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        public ViewHolder(@NonNull View itemView) {
            super(itemView);
        }

        public void onBind(FileAttachment attachment, ItemCallback callback) {
            CardView extensionContainerView = itemView.findViewById(R.id.type_indicator_view);
            TextView extensionTypeText = itemView.findViewById(R.id.type_indicator_text);
            ImageView extensionTypeImage = itemView.findViewById(R.id.type_indicator_image);

            TextView titleText = itemView.findViewById(R.id.item_title);
            TextView statusIndicator = itemView.findViewById(R.id.status_indicator);
            LinearProgressIndicator progressIndicator = itemView.findViewById(R.id.progress_indicator);

            ImageButton removeItemButton = itemView.findViewById(R.id.remove_item_button);
            removeItemButton.setOnClickListener(view -> {
                if (callback != null) callback.onRemoveItemClicked(attachment);
            });

            setProgressIndicatorState(progressIndicator, attachment.getAttachmentStatus());
            statusIndicator.setText(getStatusIndicatorText(itemView.getContext(), attachment.getAttachmentStatus()));

            ContentResolver contentResolver = itemView.getContext().getContentResolver();

            try (Cursor returnCursor = contentResolver.query(attachment.getUri(), null, null, null, null)) {
                int nameIndex = returnCursor.getColumnIndex(OpenableColumns.DISPLAY_NAME);
                int sizeIndex = returnCursor.getColumnIndex(OpenableColumns.SIZE);
                returnCursor.moveToFirst();

                String name = returnCursor.getString(nameIndex);
                long byteSize = returnCursor.getLong(sizeIndex);

                setTitleText(titleText, name, byteSize, attachment.getAttachmentStatus());

                if (isError(attachment.getAttachmentStatus())) {
                    extensionContainerView.setCardBackgroundColor(ContextCompat.getColor(itemView.getContext(), R.color.glia_system_agent_bubble_color));
                    extensionTypeImage.setVisibility(View.VISIBLE);
                    extensionTypeText.setVisibility(View.GONE);
                    extensionTypeImage.setImageResource(R.drawable.ic_info);
                    extensionTypeImage.setScaleType(ImageView.ScaleType.CENTER_INSIDE);
                } else {
                    extensionContainerView.setCardBackgroundColor(ContextCompat.getColor(itemView.getContext(), R.color.glia_brand_primary_color));
                    String mimeType = contentResolver.getType(attachment.getUri());
                    extensionTypeImage.setScaleType(ImageView.ScaleType.FIT_XY);
                    if (mimeType.startsWith("image")) {
                        extensionTypeText.setVisibility(View.GONE);
                        extensionTypeImage.setVisibility(View.VISIBLE);
                        Picasso.with(itemView.getContext()).load(attachment.getUri()).into(extensionTypeImage);
                    } else {
                        extensionTypeImage.setVisibility(View.GONE);
                        extensionTypeText.setVisibility(View.VISIBLE);
                        String extension = Utils.getExtensionByStringHandling(name).orElse("");
                        extensionTypeText.setText(extension);
                    }
                }
            }
        }

        private boolean isError(FileAttachment.Status status) {
            switch (status) {
                case ERROR_NETWORK_TIMEOUT:
                case ERROR_INTERNAL:
                case ERROR_INVALID_INPUT:
                case ERROR_PERMISSIONS_DENIED:
                case ERROR_FORMAT_UNSUPPORTED:
                case ERROR_FILE_TOO_LARGE:
                case ERROR_ENGAGEMENT_MISSING:
                case ERROR_UNKNOWN:
                    return true;
                default:
                    return false;
            }
        }

        private void setTitleText(TextView titleText, String fileName, long byteSize, FileAttachment.Status status) {
            switch (status) {
                case UPLOADING:
                case SECURITY_SCAN:
                case READY_TO_SEND:
                    titleText.setTextColor(ContextCompat.getColor(itemView.getContext(), R.color.glia_base_normal_color));
                    titleText.setText(String.format("%s • %s", fileName, Formatter.formatFileSize(itemView.getContext(), byteSize)));
                    break;
                case ERROR_NETWORK_TIMEOUT:
                    titleText.setTextColor(ContextCompat.getColor(itemView.getContext(), R.color.design_default_color_error));
                    titleText.setText(R.string.chat_attachment_upload_error_network_time_out);
                    break;
                case ERROR_INVALID_INPUT:
                    titleText.setTextColor(ContextCompat.getColor(itemView.getContext(), R.color.design_default_color_error));
                    titleText.setText(R.string.chat_attachment_upload_error_invalid_input);
                    break;
                case ERROR_PERMISSIONS_DENIED:
                    titleText.setTextColor(ContextCompat.getColor(itemView.getContext(), R.color.design_default_color_error));
                    titleText.setText(R.string.chat_attachment_upload_error_read_access_permissions_denied);
                    break;
                case ERROR_FORMAT_UNSUPPORTED:
                    titleText.setTextColor(ContextCompat.getColor(itemView.getContext(), R.color.design_default_color_error));
                    titleText.setText(R.string.chat_attachment_upload_error_file_type_invalid);
                    break;
                case ERROR_FILE_TOO_LARGE:
                    titleText.setTextColor(ContextCompat.getColor(itemView.getContext(), R.color.design_default_color_error));
                    titleText.setText(R.string.chat_attachment_upload_error_file_size_over_limit);
                    break;
                case ERROR_ENGAGEMENT_MISSING:
                    titleText.setTextColor(ContextCompat.getColor(itemView.getContext(), R.color.design_default_color_error));
                    titleText.setText(R.string.chat_attachment_upload_error_engagement_missing);
                    break;
                case ERROR_UNKNOWN:
                case ERROR_INTERNAL:
                    titleText.setTextColor(ContextCompat.getColor(itemView.getContext(), R.color.design_default_color_error));
                    titleText.setText(R.string.chat_attachment_upload_error_internal_error);
                    break;
            }
        }

        private void setProgressIndicatorState(LinearProgressIndicator progressIndicator, FileAttachment.Status status) {
            switch (status) {
                case UPLOADING:
                    progressIndicator.setIndeterminate(true);
                    progressIndicator.setProgress(0);
                    progressIndicator.setIndicatorColor(ContextCompat.getColor(itemView.getContext(), R.color.glia_brand_primary_color));
                    break;
                case SECURITY_SCAN:
                    progressIndicator.setIndeterminate(true);
                    progressIndicator.setProgress(50);
                    progressIndicator.setIndicatorColor(ContextCompat.getColor(itemView.getContext(), R.color.glia_brand_primary_color));
                    break;
                case READY_TO_SEND:
                    progressIndicator.setIndeterminate(false);
                    progressIndicator.setProgress(100);
                    progressIndicator.setIndicatorColor(ContextCompat.getColor(itemView.getContext(), R.color.glia_brand_primary_color));
                    break;
                case ERROR_NETWORK_TIMEOUT:
                case ERROR_INTERNAL:
                case ERROR_INVALID_INPUT:
                case ERROR_PERMISSIONS_DENIED:
                case ERROR_FORMAT_UNSUPPORTED:
                case ERROR_FILE_TOO_LARGE:
                case ERROR_ENGAGEMENT_MISSING:
                case ERROR_UNKNOWN:
                    progressIndicator.setIndeterminate(false);
                    progressIndicator.setProgress(100);
                    progressIndicator.setIndicatorColor(ContextCompat.getColor(itemView.getContext(), R.color.design_default_color_error));
                    break;
            }
        }

        private String getStatusIndicatorText(Context context, FileAttachment.Status status) {
            switch (status) {
                case SECURITY_SCAN:
                    return context.getString(R.string.chat_attachment_upload_checking_file);
                case READY_TO_SEND:
                    return context.getString(R.string.chat_attachment_upload_ready_to_send);
                case ERROR_NETWORK_TIMEOUT:
                case ERROR_INTERNAL:
                case ERROR_INVALID_INPUT:
                case ERROR_PERMISSIONS_DENIED:
                case ERROR_FORMAT_UNSUPPORTED:
                case ERROR_FILE_TOO_LARGE:
                case ERROR_ENGAGEMENT_MISSING:
                case ERROR_UNKNOWN:
                    return context.getString(R.string.chat_attachment_upload_failed_upload);
                case UPLOADING:
                default:
                    return context.getString(R.string.chat_attachment_upload_uploading);
            }
        }
    }

    public interface ItemCallback {
        void onRemoveItemClicked(FileAttachment attachment);
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/chat/adapter/UploadAttachmentAdapter.kt
package com.glia.widgets.chat.adapter

import android.content.Context
import android.text.format.Formatter
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.ColorInt
import androidx.annotation.StringRes
import androidx.core.view.isVisible
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import coil3.load
import com.glia.widgets.R
import com.glia.widgets.databinding.ChatAttachmentUploadedItemBinding
import com.glia.widgets.di.Dependencies
import com.glia.widgets.helper.getColorCompat
import com.glia.widgets.helper.getColorStateListCompat
import com.glia.widgets.helper.layoutInflater
import com.glia.widgets.helper.setLocaleContentDescription
import com.glia.widgets.helper.toFileExtensionOrEmpty
import com.glia.widgets.internal.fileupload.model.LocalAttachment
import com.glia.widgets.internal.fileupload.model.LocalAttachment.Status
import com.glia.widgets.locale.StringKey
import com.glia.widgets.locale.StringKeyPair
import com.glia.widgets.view.unifiedui.applyCardLayerTheme
import com.glia.widgets.view.unifiedui.applyColorTheme
import com.glia.widgets.view.unifiedui.applyImageColorTheme
import com.glia.widgets.view.unifiedui.applyTextTheme
import com.glia.widgets.view.unifiedui.theme.chat.FilePreviewTheme
import com.glia.widgets.view.unifiedui.theme.chat.FileUploadBarTheme
import com.glia.widgets.view.unifiedui.theme.chat.UploadFileTheme
import com.google.android.material.card.MaterialCardView
import com.google.android.material.progressindicator.LinearProgressIndicator
import com.google.android.material.R as Material_R

/**
 * [DiffUtil.ItemCallback] for [LocalAttachment] type
 */
internal class UploadAttachmentItemCallback : DiffUtil.ItemCallback<LocalAttachment>() {
    override fun areItemsTheSame(oldItem: LocalAttachment, newItem: LocalAttachment): Boolean =
        oldItem.uri == newItem.uri

    override fun areContentsTheSame(oldItem: LocalAttachment, newItem: LocalAttachment): Boolean =
        oldItem.isReadyToSend == newItem.isReadyToSend && oldItem.attachmentStatus == newItem.attachmentStatus
}

/**
 * Upload File Attachment Adapter
 */
internal class UploadAttachmentAdapter(private val isMessageCenter: Boolean = false) :
    ListAdapter<LocalAttachment, ViewHolder>(UploadAttachmentItemCallback()) {
    private var callback: ItemCallback? = null
    fun setItemCallback(callback: ItemCallback?) {
        this.callback = callback
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder =
        ViewHolder(
            ChatAttachmentUploadedItemBinding.inflate(parent.layoutInflater, parent, false),
            isMessageCenter
        )

    override fun onBindViewHolder(holder: ViewHolder, position: Int) =
        holder.onBind(getItem(position), callback)

    fun interface ItemCallback {
        fun onRemoveItemClicked(attachment: LocalAttachment)
    }
}

/**
 * ViewHolder
 */
internal class ViewHolder(
    private val binding: ChatAttachmentUploadedItemBinding,
    isMessageCenter: Boolean
) : RecyclerView.ViewHolder(binding.root) {

    private val extensionContainerView: MaterialCardView get() = binding.typeIndicatorView
    private val extensionTypeText: TextView get() = binding.typeIndicatorText
    private val extensionTypeImage: ImageView get() = binding.typeIndicatorImage
    private val titleText: TextView get() = binding.itemTitle
    private val statusText: TextView get() = binding.statusIndicator
    private val progressIndicator: LinearProgressIndicator get() = binding.progressIndicator
    private val removeItemButton: ImageButton get() = binding.removeItemButton

    private val context: Context get() = itemView.context
    private val localeProvider = Dependencies.localeProvider

    private val theme: FileUploadBarTheme? by lazy {
        Dependencies.gliaThemeManager.theme?.run {
            if (isMessageCenter) {
                secureMessagingWelcomeScreenTheme?.attachmentListTheme
            } else {
                chatTheme?.input?.fileUploadBar
            }
        }
    }

    private val filePreviewTheme: FilePreviewTheme? get() = theme?.filePreview

    init {
        if (isMessageCenter) {
            itemView.setBackgroundResource(R.drawable.bg_attachment)
        }
        progressIndicator.applyColorTheme(theme?.progressBackground)
        removeItemButton.applyImageColorTheme(theme?.removeButton)
        extensionTypeText.applyTextTheme(filePreviewTheme?.text, withAlignment = false)
    }

    fun onBind(attachment: LocalAttachment?, callback: UploadAttachmentAdapter.ItemCallback?) {
        val displayName = attachment?.displayName ?: return

        val size = Formatter.formatFileSize(context, attachment.size)

        removeItemButton.setOnClickListener { callback?.onRemoveItemClicked(attachment) }

        setProgressIndicatorState(attachment.attachmentStatus)

        updateTitleAndStatusText(displayName, size, attachment.attachmentStatus)

        updateExtensionType(attachment, displayName)

        updateContentDescription(displayName, size, attachment)
    }

    private fun updateExtensionType(
        attachment: LocalAttachment,
        displayName: String
    ) {
        if (attachment.attachmentStatus.isError) {
            extensionContainerView.setCardBackgroundColor(
                itemView.getColorStateListCompat(R.color.glia_neutral_color)
            )
            showExtensionTypeImage()
            extensionTypeImage.setImageResource(R.drawable.ic_info)
            extensionTypeImage.scaleType = ImageView.ScaleType.CENTER_INSIDE

            extensionContainerView.applyCardLayerTheme(filePreviewTheme?.errorBackground)
            extensionTypeImage.applyImageColorTheme(filePreviewTheme?.errorIcon)
        } else {
            extensionContainerView.setCardBackgroundColor(
                itemView.getColorCompat(R.color.glia_primary_color)
            )
            extensionTypeImage.scaleType = ImageView.ScaleType.FIT_XY
            extensionTypeImage.imageTintList = null

            if (attachment.isImage) {
                showExtensionTypeImage()
                extensionTypeImage.load(attachment.uri)
            } else {
                showExtensionTypeText()
                extensionTypeText.text = displayName.toFileExtensionOrEmpty().uppercase()
            }

            extensionContainerView.applyCardLayerTheme(filePreviewTheme?.background)
        }
    }

    private fun showExtensionTypeText() {
        extensionTypeImage.isVisible = false
        extensionTypeText.isVisible = true
    }

    private fun showExtensionTypeImage() {
        extensionTypeImage.isVisible = true
        extensionTypeText.isVisible = false
    }

    private fun updateContentDescription(
        displayName: String,
        size: String?,
        attachment: LocalAttachment
    ) {
        removeItemButton.setLocaleContentDescription(
            R.string.chat_file_remove_upload_accessibility_label,
            StringKeyPair(StringKey.NAME, displayName)
        )
        itemView.setLocaleContentDescription(
            R.string.android_chat_file_accessibility,
            StringKeyPair(StringKey.NAME, displayName),
            StringKeyPair(StringKey.SIZE, size ?: localeProvider.getRemoteString(R.string.general_unknown)),
            StringKeyPair(StringKey.STATUS, getStatusIndicatorText(attachment.attachmentStatus))
        )
    }

    private fun updateTitleAndStatusText(fileName: String, byteSize: String, status: Status) {
        val textColorRes =
            if (status.isError) Material_R.color.design_default_color_error else R.color.glia_normal_color
        titleText.setTextColor(itemView.getColorCompat(textColorRes))

        titleText.text =
            if (status.isError) localeProvider.getString(getTitleErrorStringRes(status)) else "$fileName • $byteSize"

        statusText.text = getStatusIndicatorText(status)

        val uploadFileTheme = getUploadFileTheme(status)

        titleText.applyTextTheme(
            uploadFileTheme?.text,
            withBackground = true,
            withAlignment = false
        )

        statusText.applyTextTheme(
            uploadFileTheme?.info,
            withBackground = true,
            withAlignment = false
        )
    }

    private fun getUploadFileTheme(status: Status): UploadFileTheme? = theme?.run {
        when (status) {
            Status.UPLOADING -> uploading
            Status.SECURITY_SCAN -> uploaded
            Status.READY_TO_SEND -> uploaded
            else -> error
        }
    }

    @StringRes
    private fun getTitleErrorStringRes(status: Status): Int = when (status) {
        Status.ERROR_NETWORK_TIMEOUT -> R.string.android_upload_error_network
        Status.ERROR_INVALID_INPUT -> R.string.android_upload_error_invalid_input
        Status.ERROR_PERMISSIONS_DENIED -> R.string.android_upload_error_permissions
        Status.ERROR_FORMAT_UNSUPPORTED -> R.string.chat_attachment_unsupported_file
        Status.ERROR_FILE_TOO_LARGE -> R.string.chat_file_size_limit_error
        Status.ERROR_ENGAGEMENT_MISSING -> R.string.android_upload_error_engagement_missing
        Status.ERROR_SECURITY_SCAN_FAILED -> R.string.chat_file_infected_file_error
        Status.ERROR_FILE_UPLOAD_FORBIDDEN -> R.string.android_upload_error_forbidden
        Status.ERROR_SUPPORTED_FILE_ATTACHMENT_COUNT_EXCEEDED -> R.string.android_upload_error_file_limit
        else -> R.string.error_internal
    }

    private fun setProgressIndicatorState(status: Status) {
        val normalColor = theme?.progress?.primaryColor
            ?: itemView.getColorCompat(R.color.glia_primary_color)
        val errorColor = theme?.errorProgress?.primaryColor
            ?: itemView.getColorCompat(com.google.android.material.R.color.design_default_color_error)

        when {
            status == Status.UPLOADING -> updateProgress(
                indicatorColor = normalColor,
                progressValue = 0,
                inDeterminate = true
            )

            status == Status.SECURITY_SCAN -> updateProgress(
                indicatorColor = normalColor,
                progressValue = 50,
                inDeterminate = true
            )

            status == Status.READY_TO_SEND -> updateProgress(
                indicatorColor = normalColor,
                progressValue = 100,
                inDeterminate = false
            )

            status.isError ->
                updateProgress(
                    indicatorColor = errorColor,
                    progressValue = 100,
                    inDeterminate = false
                )
        }
    }

    private fun updateProgress(
        @ColorInt indicatorColor: Int,
        progressValue: Int = 0,
        inDeterminate: Boolean = true
    ) {
        progressIndicator.apply {
            isIndeterminate = inDeterminate
            setIndicatorColor(indicatorColor)
            progress = progressValue
        }
    }

    private fun getStatusIndicatorText(status: Status): String =
        localeProvider.getString(
            when {
                status == Status.SECURITY_SCAN -> R.string.chat_file_upload_scanning
                status == Status.READY_TO_SEND -> R.string.chat_file_upload_success
                status.isError -> R.string.chat_file_upload_failed
                status == Status.UPLOADING -> R.string.chat_file_upload_in_progress
                else -> R.string.chat_file_upload_in_progress
            }
        )
}
